### PR TITLE
Release 0.5.0

### LINKED ISSUE #275: Switch to meta-QR code
As discussed in our weekly meeting, we will no longer encode the refresh token in the QR code for pairing an app. Instead we will generate a unique URL where the information for pairing will be hosted, and we will encode this URL in the QR code instead. This will make QR codes have less data encoded in them, so they become easier to scan. Additionally, related to #262, this will allow us to push more information to the client without increasing the size of the QR code.

### LINKED ISSUE #289: Admin password should be configurable via Environment variable
The admin password should be configurable via an environment variable by default unless changed via the portal. Related to https://github.com/RADAR-base/RADAR-Docker/issues/137

### LINKED ISSUE #291: Crash on radar-schemas 0.3.4
The management portal crashes when it uses catalog server with radar-schemas 0.3.4. I'm assuming this is because it does not understand the newly created `CONNECTOR` source type, and that doesn't have a vendor, model and version.

Ideally, the management portal should not crash when it encounters a wrong source type, but warn and ignore instead.

### LINKED ISSUE #292: Update logos and icons of radarbase
Update logos and icons of radarbase

### LINKED ISSUE #294: Harmonise Exception handling
Currently, there are two methods or error handling. One with HTTP-Header which is necessary for UI, and other with formatting body with message and parameters. These should be combined in ExceptionTranslator.

### LINKED ISSUE #297: Add privacy policy url to metaQR code.
Adds `privacy policy url` to metaQR code. This should give project specific url if available or a general privacy policy url configured.

### LINKED ISSUE #299: Add support to Connector type from catalog server

### LINKED ISSUE #301: Move to Staged Dockerfile

### LINKED ISSUE #304: YamlServerConfig throws NullPointerException on file not found
On line
https://github.com/RADAR-base/ManagementPortal/blob/5bca05baa55520b20a0b82d2d35ed5c1336161d5/radar-auth/src/main/java/org/radarcns/auth/config/YamlServerConfig.java#L60
if the `radar-is.yml` file is not found on the class path, the logging statement throws a `NullPointerException`. I would expect a `ConfigurationException` instead as stated in the Javadoc.

### LINKED ISSUE #308: Add RadarToken.hasPermissionOnSource method
To implement RADAR-base/RADAR-Gateway#21, it would be good if `RadarToken` exposed a `hasPermissionOnSource(Permission permission, String sourceId)` method. It should allow tokens with given permission scope and:
- from a user with Permission authority in any project and with the given `sourceId` in the `sources` claim, or
- from client credentials.

### LINKED ISSUE #311: Meta token should contain client secret too
To allow configurable platform, the MP should also send the client secret with the meta-token. Otherwise, it will add a constraint of all platforms using the same secret. This has higher risk than sending secrets with meta-token. 
FYI @blootsvoets, @dennyverbeeck , @yatharthranjan

### LINKED ISSUE #313: MP 0.4.1 fails to grant new access token for existing clients under the EC key signing alias
the MP gave the following error whenever an existing client tried to get a new access token using refresh token. Changing the signing key alias to the RSA alias and this error was gone.
```
Handling error: InvalidTokenException, Cannot convert access token to JSON

org.springframework.security.oauth2.common.exceptions.InvalidTokenException: Cannot convert access token to JSON
	at org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter.decode(JwtAccessTokenConverter.java:288)
	at org.springframework.security.oauth2.provider.token.store.JwtTokenStore.convertAccessToken(JwtTokenStore.java:88)
	at org.springframework.security.oauth2.provider.token.store.JwtTokenStore.readRefreshToken(JwtTokenStore.java:102)
	at org.springframework.security.oauth2.provider.token.DefaultTokenServices.refreshAccessToken(DefaultTokenServices.java:140)
	at org.springframework.security.oauth2.provider.refresh.RefreshTokenGranter.getAccessToken(RefreshTokenGranter.java:47)
	at org.springframework.security.oauth2.provider.token.AbstractTokenGranter.grant(AbstractTokenGranter.java:65)
	at org.springframework.security.oauth2.provider.CompositeTokenGranter.grant(CompositeTokenGranter.java:38)
	at org.springframework.security.oauth2.config.annotation.web.configurers.AuthorizationServerEndpointsConfigurer$4.grant(AuthorizationServerEndpointsConfigurer.java:561)
	at org.springframework.security.oauth2.provider.endpoint.TokenEndpoint.postAccessToken(TokenEndpoint.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at com.codahale.metrics.servlet.AbstractInstrumentedFilter.doFilter(AbstractInstrumentedFilter.java:111)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:96)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:110)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilterInternal(BasicAuthenticationFilter.java:215)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:108)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:81)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:64)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:332)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.radarcns.management.security.jwt.SignatureException: An error occured verifying the signature
	at org.radarcns.management.security.jwt.EcdsaVerifier.verify(EcdsaVerifier.java:35)
	at org.springframework.security.jwt.JwtImpl.verifySignature(JwtHelper.java:287)
	at org.springframework.security.jwt.JwtHelper.decodeAndVerify(JwtHelper.java:77)
	at org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter.decode(JwtAccessTokenConverter.java:277)
	... 121 common frames omitted
Caused by: java.security.SignatureException: Could not verify signature
	at sun.security.ec.ECDSASignature.engineVerify(ECDSASignature.java:325)
	at java.security.Signature$Delegate.engineVerify(Signature.java:1222)
	at java.security.Signature.verify(Signature.java:655)
	at org.radarcns.management.security.jwt.EcdsaVerifier.verify(EcdsaVerifier.java:31)
	... 124 common frames omitted
Caused by: java.security.SignatureException: Invalid encoding for signature
	at sun.security.ec.ECDSASignature.decodeSignature(ECDSASignature.java:400)
	at sun.security.ec.ECDSASignature.engineVerify(ECDSASignature.java:322)
	... 127 common frames omitted
Caused by: java.io.IOException: Sequence tag error
	at sun.security.util.DerInputStream.getSequence(DerInputStream.java:330)
	at sun.security.ec.ECDSASignature.decodeSignature(ECDSASignature.java:376)
	... 128 common frames omitted
```

### LINKED ISSUE #317: Set OAuth client secret to empty for public OAuth clients
Use empty OAuth client secret, such as described in [this comment](https://github.com/RADAR-base/ManagementPortal/pull/312#issuecomment-415496716). For more info see discussion in #311.

### PR DESCRIPTION
**Pull request by nivemaham**
# Changes from release 0.4.1
**MetaQR code** #275 
- Adds a new entity to store refresh-token for subjects and relevant meta-data.
- Send tokenUrl in oauth-pair response
- Add new resource to fetch meta-token
- Add privacy-policy-url and baseUrl to metaToken #297
- ~~Share client secret with meta-token~~ #311  (Client secret no longer used for public OAuth clients, see links to discussions in #317)

**Exception Harmonization** #294 
- We have new exceptions based on HTTP Status code conventions and a ExceptionTranslator to implement body and headers.

**Other fixes**
- Ignore exceptions during SourceTypeLoading #291 
- Add support for `CONNECTOR` types in SourceTypeLoader #299
- Configurable `admin` password #289
- Upgrade `lodash` dependencies.
- Fix `NullPointerException` during configLoading. #304
- Update logos and icons of `radar-base.org`. #292 
- Add `hasPermissionOnSource` #308 
- Dockerfile build staging (#301)
- Update to the JWT signing code (#315)
- Fix OAuth environment variable documentation
- Allow public OAuth clients to use `authorization_code` and `refresh_token` flows without the use of a client secret (#317).

Closes #275, closes #289, closes #291, closes #292, closes #294, closes #297, closes #299, closes #301, closes #304, closes #308, closes #311, closes #313, closes #317.

### CODE REVIEW

File: src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     * @return the client as a {@link ClientPairInfoDTO}
+     */
+    @GetMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
+    @Timed
+    public ResponseEntity<TokenDTO> getTokenByTokenName(@PathVariable("tokenName") String tokenName)
+            throws MalformedURLException {
+        log.info("Requesting token with tokenName {}", tokenName);
+        try {
+            // sleep for 5 seconds
+            TimeUnit.SECONDS.sleep(5);
```
Comment (dennyverbeeck) [inline src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java]: Why do we need to `sleep()` here?

File: src/main/java/org/radarcns/management/domain/MetaToken.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    @JoinColumn(name = "subject_id")
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    private Subject subject;
+
+    /**
+     * Meta token constructor.
+     * Generates a random string as the tokenName.
+     */
+    public MetaToken() {
+        if (this.tokenName == null) {
```
Comment (dennyverbeeck) [inline src/main/java/org/radarcns/management/domain/MetaToken.java]: Isn't the `tokenName` always `null` at the time this constructor is called?

File: src/main/java/org/radarcns/management/domain/MetaToken.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    private Instant expiryDate = null;
+
+    @ManyToOne(fetch = FetchType.EAGER)
+    @JoinColumn(name = "subject_id")
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+    private Subject subject;
+
+    /**
+     * Meta token constructor.
+     * Generates a random string as the tokenName.
```
Comment (dennyverbeeck) [inline src/main/java/org/radarcns/management/domain/MetaToken.java]: Perhaps we should rephrase this as something like 'Must generate a random string as the tokenName'. Since other parts in the code rely on the fact that this is happening.

Comment (dennyverbeeck) [reply src/main/java/org/radarcns/management/domain/MetaToken.java]: We can even add a `MetaTokenUnitTest.java` to check that a random `tokenName` is indeed generated. I'm just a bit concerned that if at some point someone removes the generation of `tokenName` for some reason, it will break `MetaTokenService.saveUniqueToken()`, which will not be immediately apparent otherwise.

File: src/main/java/org/radarcns/management/service/MetaTokenService.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+     * @return fetched token as {@link MetaToken}.
+     */
+    @Transactional(readOnly = true)
+    public MetaToken getToken(String tokenName) {
+        Optional<MetaToken> fetchedToken = metaTokenRepository.findOneByTokenName(tokenName);
+
+        if (fetchedToken.isPresent()) {
+            return  fetchedToken.get();
+        } else {
+            throw  new NotFoundException("Meta token not found with tokenName", META_TOKEN,
```
Comment (dennyverbeeck) [inline src/main/java/org/radarcns/management/service/MetaTokenService.java]: There are two spaces between `throw` and `new`, and between `return` and `fetchedToken.get()` 😛

File: src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java (hunk 1 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java]: Perhaps 5 seconds is excessive (3 seconds would probably do), but it makes brute-force attacks too expensive to be worthwhile. For every guess, an attacker needs to wait X seconds before a response is returned. This needs to be done after both valid and invalid requests, to ensure that an attacker does not define an internal time threshold that would only be reached for invalid URLs.

Attack pseudo-code:
```python
for sequence in sequences_to_try():
     response = request(sequence, timeout_ms=500)
     if response.successful:
         # keys to the castle!
```

Perhaps it would be good to mention this in the comments, indeed.

**Comment by blootsvoets** [comment]
@dennyverbeeck in other RADAR projects, tests are categorised based on gradle `sourceSet`. That way a test is run depending on the directory it is placed in (e.g. `src/test`, `src/integrationTest`, `src/cucumberTest`) and the class name is inconsequential. Would this help for future such errors?

File: src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java (hunk 1 above)
Comment (dennyverbeeck) [reply src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java]: I checked with gatling to be sure, and it is as i suspected. This does indeed make each request take a long time, but requests are served asynchronously. So an attacker can just make 100's of requests per second, and would receive all the responses a few seconds later.

See this image, where we can see around 15 requests per second, and also 15 responses per second, as opposed to one response per 5 seconds. Generated using [this](https://gist.github.com/dennyverbeeck/79836a3a529154ebfca6d38268e2d5ba) code.
image

We will probably need to add an interceptor to implement rate limiting based on the client's IP address. I suggest we remove the sleep and move the rate limiting implementation to a new issue.

Comment (blootsvoets) [reply src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java]: There is a limit in the number of simultaneous connections for clients though, which makes the sleep more effective. I agree that a rate limiting implementation would be preferable, but I would rather wait for that implementation and then remove the sleep. In the context of RADAR, perhaps the rate limiting needs to happen on the nginx level, since that is the reverse proxy on top of MP.

Comment (nivemaham) [reply src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java]: I agree to remove the sleep after we have the rate-limiting. IMO, having the sleep doesn't do any harm.
